### PR TITLE
KTOR-5199 Support WebSockets in Curl engine

### PR DESCRIPTION
**Pull request by dtretyakov**
**Subsystem**
Client, Curl engine

**Motivation**
[Curl 7.86.0](https://curl.se/changes.html#7_86_0) added experimental support for WebSockets.

**Solution**
This PR brings support of experimental WebSockets in libcurl [KTOR-5199](https://youtrack.jetbrains.com/issue/KTOR-5199/Support-WebSockets-in-Curl-engine).

To verify WebSockets availability we're using `curl_version_info` which returns list of enabled protocols.

The `CurlResponseBodyData` become interface with `CurlHttpResponseBody`/`CurlWebSocketResponseBody` implementations.

The `bodyStartedReceiving` is used to detect the end of headers section.

We're using WebSocket with callbacks approach where:
* `curl_ws_send` used to send outgoing frames
* `onBodyChunkReceived` with `curl_ws_meta` used to receive incoming frames

**Environment**
Since WebSockets feature is experimental we need to enable them during curl compliation.

*macOS*
```
curl https://raw.githubusercontent.com/dtretyakov/homebrew-cloudflare/master/curl.rb -o curl.rb
brew install -s curl.rb 
```

*linux*
* [Reference](https://stackoverflow.com/a/76769965)
* [Commit](https://github.com/curl/curl/commit/6c429d4225b946116eadcd4d575b267a0aad2ecd)
* [Workflow file](https://github.com/curl/curl/actions/runs/7618741488/workflow)

### CODE REVIEW

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlWebSocketSession.kt [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+) : WebSocketSession, Closeable {
+
+    private val closed = atomic(false)
+    private val socketJob = Job(callContext[Job])
+    private val _outgoing = Channel<Frame>(Channel.UNLIMITED)
+
+    override val coroutineContext: CoroutineContext = callContext + socketJob + CoroutineName("curl-ws")
+    override var masking: Boolean
+        get() = true
+        set(_) {}
```
Comment (bjhham) [inline ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlWebSocketSession.kt:29]: This unimplemented stuff makes me think there's too many things in the `WebSocketSession` interface... I see there's other implementations with similar declarations.

File: ktor-client/ktor-client-curl/desktop/test/io/ktor/client/engine/curl/test/CurlWebSocketTests.kt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+            client.webSocket("$TEST_WEBSOCKET_SERVER/websockets/echo") {
+                send(Frame.Text(""))
+
+                val actual = incoming.receive()
+
+                assertTrue(actual is Frame.Text)
+                assertEquals("", actual.readText())
+            }
+        }
+    }
```
Comment (bjhham) [inline ktor-client/ktor-client-curl/desktop/test/io/ktor/client/engine/curl/test/CurlWebSocketTests.kt:52]: Would be good to add tests for other frame types, sending multiple frames.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt [hunk 3]
```diff
@@ -42,6 +43,11 @@ internal class CurlProcessor(coroutineContext: CoroutineContext) {
     }
 
     suspend fun executeRequest(request: CurlRequestData): CurlSuccess {
+        if (request.isUpgradeRequest && !curlProtocols.contains("ws")) {
+            @Suppress("DEPRECATION")
+            throw CurlRuntimeException("WebSockets are supported in experimental libcurl 7.86 and greater")
```
Comment (e5l) [inline ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt:48]: please avoid using deprecated exception

**Review by e5l — CHANGES_REQUESTED** [review]
Hey @dtretyakov, thanks for the PR.

Please let me know when you have an update about the Curl binary version. I will run tests on CI

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt [hunk 4]
```diff
@@ -42,6 +43,11 @@ internal class CurlProcessor(coroutineContext: CoroutineContext) {
...
+        }
+
```
[the remark this answers is not in the document]
Comment (DRSchlaubi) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt:50]: It wouldn't since `curlProtocols` is a list of protocols supported by this curl installation, so if `wss` is in there `ws` would be in there to, therefore passing this check, although it could probabbly be optimized to be

```suggestion
        if (request.isUpgradeRequest && !curlProtocols.contains(request.url.substringBefore("://"))) {
            @Suppress("DEPRECATION")
            throw CurlRuntimeException("WebSockets are supported in experimental libcurl 7.86 and greater")
        }

```

**Comment by DRSchlaubi** [comment]
Trying to connect to `wss://gateway.discord.gg/?v=10&encoding=json&compress=zlib-stream` is returning `400 Bad request` so there are some issues with this implementation`

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlClientEngine.kt [hunk 5]
```diff
@@ -46,9 +46,15 @@ internal class CurlClientEngine(
             rawHeaders.release()
 
             val responseBody: Any = if (needToProcessSSE(data, status, headers)) {
-                DefaultClientSSESession(data.body as SSEClientContent, bodyChannel, callContext)
+                val content = data.body as SSEClientContent
+                val body = responseBody as CurlHttpResponseBody
+                DefaultClientSSESession(content, body.bodyChannel, callContext)
+            } else if (data.isUpgradeRequest()) {
+                val websocket = responseBody as CurlWebSocketResponseBody
+                CurlWebSocketSession(websocket, callContext)
```
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlClientEngine.kt:54]: Actually WebSocket extension [checks protocol upgrade](https://github.com/ktorio/ktor/blob/830510d428a02cdf785d3e53d0a5fb2d52431fed/ktor-client/ktor-client-core/common/src/io/ktor/client/plugins/websocket/WebSockets.kt#L181) and will throw like that:
```
io.ktor.client.plugins.websocket.WebSocketException: Handshake exception, expected status code 101 but was 400
```

**Review by DRSchlaubi — COMMENTED** [review]
```suggestion
    if (isUpgradeRequest()) {
        // > If you add a header with no content as in 'Accept:' (no data on the right side of the colon),
        // > the internally used header is disabled/removed.
        // https://curl.se/libcurl/c/CURLOPT_HTTPHEADER.html
        // These need to be removed as they're set by Ktor in WebSocketContent
        // Leaving them can cause an HTTP 400 on some servers, as they're present twice

        curl_slist_append(result, "Sec-WebSocket-Key: ")
        curl_slist_append(result, "Sec-WebSocket-Version: ")
    }
```

**Comment by DRSchlaubi** [comment]
Possible fix for duplicated header issue

```diff
Subject: [PATCH] Fix duplicate headers
---
Index: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt b/ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt
--- a/ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt	(revision 6265a80481d77cdbb2ff7950750e8ee5aa085813)
+++ b/ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt	(date 1706196036063)
@@ -71,8 +71,11 @@
     var result: CPointer<curl_slist>? = null
 
     mergeHeaders(headers, body) { key, value ->
-        val header = "$key: $value"
-        result = curl_slist_append(result, header)
+        // These are set by curl itself
+        if (key != "Sec-WebSocket-Key" && key != "Sec-WebSocket-Version") {
+            val header = "$key: $value"
+            result = curl_slist_append(result, header)
+        }
     }
 
     result = curl_slist_append(result, "Expect:")

```

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt]: Thanks, I added the actual protocol verification.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt [hunk 6]
```diff
@@ -87,3 +87,19 @@ internal fun UInt.fromCurl(): HttpProtocolVersion = when (this) {
     /* old curl fallback */
     else -> HttpProtocolVersion.HTTP_1_1
 }
+
+@OptIn(ExperimentalForeignApi::class)
+internal fun getCurlProtocols(): List<String> {
```
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt:110]: Added a comment what this protocols list represents.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt [hunk 7]
```diff
@@ -87,3 +87,19 @@ internal fun UInt.fromCurl(): HttpProtocolVersion = when (this) {
...
+    val protocols = versionInfo.protocols ?: return emptyList()
+    var index = 0
+
+    return buildList {
+        do {
+            val protocol = protocols[index++] ?: break
+            add(protocol.toKString())
+        } while (true)
+    }
+}
```
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt]: Extracted a helper method to convert null terminated characters array to Kotlin strings.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlCallbacks.kt [hunk 8]
```diff
@@ -20,9 +19,15 @@ internal fun onHeadersReceived(
...
-    val packet = userdata.fromCPointer<CurlResponseBuilder>().headersBytes
+    val response = userdata.fromCPointer<CurlResponseBuilder>()
+    val packet = response.headersBytes
     val chunkSize = (size * count).toLong()
     packet.writeFully(buffer, 0, chunkSize)
+
+    if (chunkSize == 2L && buffer[0] == 0x0D.toByte() && buffer[1] == 0x0A.toByte()) {
+        response.bodyStartedReceiving.complete(Unit)
+    }
+
```
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlCallbacks.kt]: Extracted a helper with meaningful name and comments.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlHttpResponseBody.kt [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            runBlocking {
+                bodyChannel.writeFully(buffer, 0, chunkSize)
+            }
+            chunkSize
+        } catch (cause: Throwable) {
+            return -1
+        }
+        if (written > 0) {
+            bytesWritten += written
+        }
```
[the remark this answers is not in the document]
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlHttpResponseBody.kt:43]: I'm for making these changes separaterly since as of now it was moved as is from the `onBodyChunkReceived` callback.

File: ktor-client/ktor-client-curl/desktop/test/io/ktor/client/engine/curl/test/CurlWebSocketTests.kt (hunk 2 above)
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/test/io/ktor/client/engine/curl/test/CurlWebSocketTests.kt:53]: As I understand we have the following policy:
* Common tests are placed inside `:ktor-client:ktor-client-tests`
* Engine specific for corner cases as stored inside corresponding module.

It seems that the [`WebSocketTest`](https://github.com/ktorio/ktor/blob/a1d639cc38e7bd1becb3e27cc31451d55a4837ab/ktor-client/ktor-client-tests/common/test/io/ktor/client/tests/WebSocketTest.kt#L25) covers the mentioned type of data ^.

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt (hunk 3 above)
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/CurlProcessor.kt]: replaced by the `error(...)`

**Comment by dtretyakov** [comment]
@DRSchlaubi thanks a lot for verification and review.

> New issue `curl_multi_perform` seems to block until the connection is done, making any additional requests impossible

Maybe you could share a test case when it could be reproduced?

> Trying to connect to `wss://gateway.discord.gg/?v=10&encoding=json&compress=zlib-stream`
> Possible fix for duplicated header issue

Thanks, I added a list of WebSocket headers which are handled by cURL itself, so connection to this service should work.

**Comment by DRSchlaubi** [comment]
> Thanks, I added a list of WebSocket headers which are handled by cURL itself, so connection to this service should work.
Where did you do that? Am I just blind?

> Maybe you could share a test case when it could be reproduced?

Sure

```kotlin
fun main() = runBlocking { 
    val client = HttpClient { 
        install(WebSockets)
    }
    
    launch { 
        client.webSocket("wss://echo.websocket.org") {
            while(!incoming.isClosedForReceive) {
              // Keep connection alive
                incoming.receive()
            }
        }
    }
    
    println("Wait for websocket to connect")
    delay(10.seconds)
    println("Requesting now!")
    
    println(client.get("https://httpbin.org/200"))
}
```

**Comment by DRSchlaubi** [comment]
After the latest changes most headers are not set, including "Authorization", which is not great

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt [hunk 10]
```diff
@@ -71,8 +82,10 @@ internal fun HttpRequestData.headersToCurl(): CPointer<curl_slist> {
     var result: CPointer<curl_slist>? = null
 
     mergeHeaders(headers, body) { key, value ->
-        val header = "$key: $value"
-        result = curl_slist_append(result, header)
+        if (isUpgradeRequest() && !DISALLOWED_WEBSOCKET_HEADERS.contains(key)) {
```
Comment (DRSchlaubi) [inline ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt:85]:
```suggestion
        if (!isUpgradeRequest() || !DISALLOWED_WEBSOCKET_HEADERS.contains(key)) {
```

Comment (DRSchlaubi) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt:85]: Otherwise normal requests don't set headers at all

**Comment by DRSchlaubi** [comment]
After more debugging it seems like the issue is this loop
```kotlin
    @OptIn(ExperimentalForeignApi::class)
    internal fun perform() {
        if (activeHandles.isEmpty()) return

        memScoped {
            val transfersRunning = alloc<IntVar>()
            do {
                println("Transfers running:${transfersRunning.value}")
                synchronized(easyHandlesToUnpauseLock) {
                    var handle = easyHandlesToUnpause.removeFirstOrNull()
                    while (handle != null) {
                        curl_easy_pause(handle, CURLPAUSE_CONT)
                        handle = easyHandlesToUnpause.removeFirstOrNull()
                    }
                }
                println("Running curl_multi_perform")
                curl_multi_perform(multiHandle, transfersRunning.ptr).verify()
                println("Ran curl_multi_perform")
                if (transfersRunning.value != 0) {
                    curl_multi_poll(multiHandle, null, 0.toUInt(), 10000, null).verify()
                    println("Ran curl_multi_poll")
                }
                if (transfersRunning.value < activeHandles.size) {
                    handleCompleted()
                    println("Ran complete")
                }
                println("Loop done")
            } while (transfersRunning.value != 0)
        }
    }
    ```
    
    Because `while (transfersRunning.value != 0)` will remain true as long as the ws connection is still active

File: ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt (hunk 10 above)
Comment (dtretyakov) [reply ktor-client/ktor-client-curl/desktop/src/io/ktor/client/engine/curl/internal/CurlAdapters.kt]: Thanks, it was addressed in the recent changes.

**Comment by dtretyakov** [comment]
> After more debugging it seems like the issue is this loop
> Because `while (transfersRunning.value != 0)` will remain true as long as the ws connection is still active

@DRSchlaubi hopefully it was also addressed in the last commit, so you could try updating.

**Comment by DRSchlaubi** [comment]
At first glance your new test case should cover this, will test myself tmr :+1:

**Comment by DRSchlaubi** [comment]
That indeed fixed it

**Comment by dtretyakov** [comment]
> Please let me know when you have an update about the Curl binary version. I will run tests on CI

Currently we're waiting when static linking of libcurl with it's dependencies will be in the repository for all targets and after that could merge the changes.

* Prototype: https://github.com/dtretyakov/ktor/commit/3c4dcd678990f2c7054bc0d6c3c4c92b49202de5
* Build scripts which could be used to create static libraries for linux/macOS: https://github.com/stunnel/static-curl

**Comment by DRSchlaubi** [comment]
Any updates

**Comment by git-bruh** [comment]
Hey @dtretyakov , thanks a lot for your work on this PR. I noticed some strange behaviour when receiving payloads greater than 1024 bytes and it seems to be a libcurl quirk, have raised a PR here: https://github.com/dtretyakov/ktor/pull/1

**Review by e5l — APPROVED** [review]
Hey, @dtretyakov! Sorry for the long delay.
Could you rebase it on `3.1.0-eap` branch before merging?

**Comment by DRSchlaubi** [comment]
I pushed my work to [curl/ws](https://github.com/kordlib/ktor/tree/curl/ws)

**Comment by dtretyakov** [comment]
> Could you rebase it on `3.1.0-eap` branch before merging?

It was rebased, but now we need to have a libcurl with websockets enabled.

**Comment by dtretyakov** [comment]
It looks that Conan libcurl receipt does no support option to enable WebSockets: https://github.com/conan-io/conan-center-index/issues/26221

**Comment by osipxd** [comment]
@dtretyakov, could you please re-open this PR from `main`? It was automatically closed by GitHub after merging of `3.1.0-eap` branch.

**Comment by whyoleg** [comment]
@osipxd, @dtretyakov JIC, I've reopened PR, so that may be it will be possible to squeeze it in 3.1.0 as we now have curl build there with WS enabled :)
https://github.com/ktorio/ktor/pull/4656
